File: l33tcode/median-of-two-sorted-arrays.py
MIN, MAX = -(10**6) - 1, 10**6 + 1


class Solution:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        first = nums2 if len(nums1) >= len(nums2) else nums1
        second = nums1 if len(nums1) >= len(nums2) else nums2

        left, right = 0, len(first)
        median = (len(first) + len(second) + 1) // 2

        while left <= right:
            mid_first = left + (right - left) // 2
            mid_second = median - mid_first

            first_before = first[mid_first - 1] if mid_first > 0 else MIN
            first_after = first[mid_first] if mid_first < len(first) else MAX
            second_before = second[mid_second - 1] if mid_second > 0 else MIN
            second_after = second[mid_second] if mid_second < len(second) else MAX

            if first_before <= second_after and second_before <= first_after:
                if (len(first) + len(second)) % 2 == 0:
                    return (
                        max(first_before, second_before)
                        + min(first_after, second_after)
                    ) / 2.0
                else:
                    return float(max(first_before, second_before))
            elif first_before > second_after:
                right = mid_first - 1
            else:
                left = mid_first + 1

        raise ValueError("Median doesn't exist")

File: l33tcode_rs/src/median_of_two_sorted_arrays.rs
#[allow(dead_code)]
struct Solution;

impl Solution {
    #[allow(dead_code, clippy::needless_pass_by_value)]
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        const MIN: i32 = -10i32.pow(6) - 1;
        const MAX: i32 = 10i32.pow(6) + 1;
        let mut nums_larger = &nums1;
        let mut nums_smaller = &nums2;
        if nums1.len() < nums2.len() {
            nums_larger = &nums2;
            nums_smaller = &nums1;
        }
        let (mut left, mut right) = (0, nums_smaller.len());

        let median = (nums_smaller.len() + nums_larger.len()).div_ceil(2);

        while left <= right {
            let mid_smaller = left + (right - left) / 2;
            let mid_larger = median - mid_smaller;

            let larger_before = nums_larger.get(mid_larger - 1).unwrap_or(&MIN);
            let larger_after = nums_larger.get(mid_larger).unwrap_or(&MAX);

            let smaller_before = nums_smaller.get(mid_smaller - 1).unwrap_or(&MIN);
            let smaller_after = nums_smaller.get(mid_smaller).unwrap_or(&MAX);

            if larger_before <= smaller_after && smaller_before <= larger_after {
                if (nums_larger.len() + nums_smaller.len()) % 2 == 0 {
                    return f64::from(
                        larger_before.max(smaller_before) + larger_after.min(smaller_after),
                    ) / 2.0;
                }
                return f64::from(*larger_before.max(smaller_before));
            } else if smaller_before > larger_after {
                right = mid_smaller;
            } else {
                left = mid_smaller + 1;
            }
        }

        panic!("Median doesn't exist")
    }
}
